### ISSUE TITLE
LDAP SSL with invalid CA

### ISSUE THREAD

**Issue by david-belbeze** [cannatag/ldap3#912]
Hello,

I tried to use the ldap3 library to connect to a serveur that use a SSL certificate with an invalid CA (probably self signed certificate).

So I used the code bellow

```python
import ssl

from ldap3 import Server, Connection, ALL, Tls


server = Server('ldaps://DNS:PORT', tls=Tls(validate=ssl.CERT_NONE))

conn = Connection(server, user="...", password="...", authentication='SIMPLE')

conn.open()
# ...
```

When I execute the code, Python print the next stack trace:

```txt
Traceback (most recent call last):
  File "C:/project\main.py", line 37, in <module>
    conn.open()
  File "C:\project\venv\lib\site-packages\ldap3\strategy\sync.py", line 57, in open
    BaseStrategy.open(self, reset_usage, read_server_info)
  File "C:\project\venv\lib\site-packages\ldap3\strategy\base.py", line 146, in open
    raise exception_history[0][0]
ldap3.core.exceptions.LDAPSocketOpenError: socket ssl wrapping error: [Errno 2] No such file or directory
```

I executed the code using the debugger and I think that the problem come from the next code:

```python
# ldap3/strategy/base.py (l.270)
        if use_ssl:
            try:
                self.connection.server.tls.wrap_socket(self.connection, do_handshake=True)
                if self.connection.usage:
                    self.connection._usage.wrapped_sockets += 1
            except Exception as e:
                self.connection.last_error = 'socket ssl wrapping error: ' + str(e)
                if log_enabled(ERROR):
                    log(ERROR, '<%s> for <%s>', self.connection.last_error, self.connection)
                raise communication_exception_factory(LDAPSocketOpenError, type(e)(str(e)))(self.connection.last_error)
```

The **do_handshake=True** is probably the problem.

I tried the low level connection to the ldap server using `socket` and `ssl` python module:

```python
import socket
import ssl

PORT =

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(('DNS', PORT))

print(s)

# Create ssl context and wrap the socket
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

# do_handshake_on_connect=do_handshake in ldap3/core/tls.py l.204
ss = ctx.wrap_socket(s, server_side=False, do_handshake_on_connect=True)
print(ss)
```

This produce the same "Errno 2" error, and when I tried **do_handshake_on_connect=False** the SSLsocket is correctly created.

ldap3 version: **2.9**
OS: **Win10**
Python: **3.8**

**Comment by zorn96** [comment]
hi @david-belbeze ! I think that this might be an issue with how your OS and python are interacting around SSL.
the "do_handshake_on_connect" is simply a part of the LDAPS protocol; when a client use LDAPS, all opened connections should immediately perform a TLS handshake.
performing the handshake later is called "opportunistic TLS", and would use a plain "ldap://" uri and the startTLS operation (start_tls()) later

but in the case where you manually made a socket - SSL hasn't actually been setup. if you don't do the handshake on connect, then you must do it later or your traffic won't be secured by TLS.
it seems like the python SSL library is encountering an error when trying to negotiate TLS on your machine.

from some searching online, there can be a few causes of this. one cause is apparently that if the server and your client cannot agree on any TLS ciphers, this is the error python raises (apparently people have run into this issue with self-signed certs a lot). another potential source of error is that python has default locations for CA certs and such, and if those are missing and no other location is specified then this can happen

can your client connect to other machines using TLS from python? or is it just this server that's the problem?
a packet capture/tcpdump will let you know if the TLS handshake results in no mutually agreed upon ciphers between the client. tracing the stacktrace further into the SSL library will let you know if you're missing files locally

let me know if what I wrote makes sense/if I can clarify anything, and what you find :)

**Comment by david-belbeze** [comment]
Ok,

Thank you, you help me with some reading to understand how TLS works.

To start, I tried the same code on Linux device and same Python version, but it raise an error, Errno 0 ...

So I continue to test the low level socket connection with ssl:
 - With the same version of Python or Python 3.7, it results in OSError [Errno 0]
 - With Python 2 or Python 3.6, I can correctly establish the connection

In debug mode, I tried to follow the execution in ssl module but debugger stopped when `self._sslobj.do_handshake` is called (probably C code) and the error is raised. So I cannot determine the problem in ssl handshake.
Next, I tried to connect to https serveur that use a certificate signed by the same CA and use the same TLS version and it works correctly with Python 3.8.
And finally I tried another ldap serveur that expose a similar certificate than the first LDAP serveur and it produce an Errno 2 on Win10 and Errno 0 on Linux.

So I think that it is more a ssl problem with new Python version (3.9 not tested) than ldap3 problem.
